waku/v2/peermanager/topic_event_handler_test.go

```diff
@@ -0,0 +1,213 @@
+	log := utils.Logger()
+	pubSubTopic := "/waku/2/go/pm/test"
+	ctx := context.Background()
+
+	// Relay and Host
+	r, h1, _ := makeWakuRelay(t, log)
+	err := r.Start(ctx)
+	require.NoError(t, err)
+
+	// Peer manager with event bus
+	pm, _ := makePeerManagerWithEventBus(t, r, &h1)
+	pm.ctx = ctx
+
+	// Start event loop to listen to events
+	ctxEventLoop, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	go pm.peerEventLoop(ctxEventLoop)
+
+	// Subscribe to Pubsub topic
+	_, err = r.Subscribe(ctx, protocol.NewContentFilter(pubSubTopic))
+	require.NoError(t, err)
+
+	// Wait for event loop to call handler
+	time.Sleep(200 * time.Millisecond)
+
+	// Check Peer Manager knows about the topic
+	pm.topicMutex.RLock()
+	_, ok := pm.subRelayTopics[pubSubTopic]
+	require.True(t, ok)
+	pm.topicMutex.RUnlock()
+
+	// UnSubscribe from Pubsub topic
+	err = r.Unsubscribe(ctx, protocol.NewContentFilter(pubSubTopic))
+	require.NoError(t, err)
+
+	// Wait for event loop to call handler
+	time.Sleep(200 * time.Millisecond)
+
+	// Check the original topic was removed from Peer Manager
+	pm.topicMutex.RLock()
+	_, ok = pm.subRelayTopics[pubSubTopic]
+	require.False(t, ok)
+	pm.topicMutex.RUnlock()
+
+	r.Stop()
+}
+
+func TestHandlePeerTopicEvent(t *testing.T) {
+	log := utils.Logger()
+	pubSubTopic := "/waku/2/go/pm/test"
+	ctx := context.Background()
+
+	hosts := make([]host.Host, 5)
+	relays := make([]*relay.WakuRelay, 5)
+
+	for i := 0; i < 5; i++ {
+		relays[i], hosts[i], _ = makeWakuRelay(t, log)
+		err := relays[i].Start(ctx)
+		require.NoError(t, err)
+	}
+
+	// Create peer manager instance with the first hosts
+	pm, eventBus := makePeerManagerWithEventBus(t, relays[0], &hosts[0])
+	pm.ctx = ctx
+	pm.RegisterWakuProtocol(relay.WakuRelayID_v200, relay.WakuRelayENRField)
+
+	// Connect host[0] with all other hosts to reach 4 connections
+	for i := 1; i < 5; i++ {
+		pm.host.Peerstore().AddAddrs(hosts[i].ID(), hosts[i].Addrs(), peerstore.PermanentAddrTTL)
+		err := pm.host.Connect(ctx, hosts[i].Peerstore().PeerInfo(hosts[i].ID()))
+		require.NoError(t, err)
+		err = pm.host.Peerstore().(wps.WakuPeerstore).SetDirection(hosts[i].ID(), network.DirOutbound)
+		require.NoError(t, err)
+
+	}
+
+	// Wait for connections to settle
+	time.Sleep(2 * time.Second)
+
+	if len(pm.host.Peerstore().(*wps.WakuPeerstoreImpl).PeersByPubSubTopic(pubSubTopic)) == 0 {
+		log.Info("No peers for the topic yet")
+	}
+
+	// Subscribe to Pubsub topic on first host only
+	_, err := relays[0].Subscribe(ctx, protocol.NewContentFilter(pubSubTopic))
+	require.NoError(t, err)
+
+	// Start event loop to listen to events
+	ctxEventLoop := context.Background()
+	go pm.peerEventLoop(ctxEventLoop)
+
+	// Prepare emitter
+	emitter, err := eventBus.Emitter(new(relay.EvtPeerTopic))
+	require.NoError(t, err)
+
+	// Send PEER_JOINED events for hosts 2-5
+	for i := 1; i < 5; i++ {
+		err = emitTopicEvent(pubSubTopic, hosts[i].ID(), emitter, relay.PEER_JOINED)
```

Comment:
As i had mentioned before, there is no need to emit these events.
Just subscribe all the 5 hosts to same pubsubTopic and these events would automatically get generated.

Reply:
That could probably be happening because the eventLoop was started after subscribe.

Reply:
I've moved eventLoop() goroutine start ahead of subscription. https://github.com/waku-org/go-waku/pull/1035/commits/0372719e4713f5969ded5060c4974acb7460be97

Result is the same. I am not getting PEER_JOINED, relay.PEER_LEFT. There is only one place in code where those get emited: (w *WakuRelay) topicEventPoll(topic string, handler *pubsub.TopicEventHandler) which doesn't use (pm *PeerManager) handlerPeerTopicEvent(peerEvt relay.EvtPeerTopic). It looks we expect something to happen at pubsub (libp2p) level and then react on our (peermanager) side to it. I was not able to find what is that specifically. Your thoughts ?

Reply:
Interesting. 
One thing i had noticed is that relay[0] was not subscribed to pubsubTopic in the code. 
But even after that events are not getting generated. Not sure what could be causing the issue.

Reply:
How should we proceed with this PR ? I have a milestone to finish this week. I really like to crack mysteries, right now though I suggest:
1) Open new issue to look at the libp2p/pm event transitions later  
2) I revert test code back to test pm's side of the event loop only

What do you think ?

Reply:
OK, we can take that approach.